cmd/gardenlet/app/app.go

```diff
@@ -557,6 +568,83 @@ func (g *garden) cleanupStaleShootStates(ctx context.Context, gardenClient clien
 	return flow.Parallel(taskFns...)(ctx)
 }
 
+func (g *garden) cleanupStaleAddonsMR(ctx context.Context, gardenClient, seedClient client.Client) error {
+	shootNamespaceList := &corev1.NamespaceList{}
+	if err := seedClient.List(ctx, shootNamespaceList, client.MatchingLabels{v1beta1constants.GardenRole: v1beta1constants.GardenRoleShoot}); err != nil {
+		return err
+	}
+
+	var taskFns []flow.TaskFn
+
+	for _, ns := range shootNamespaceList.Items {
+		namespace := ns
+
+		taskFns = append(taskFns, func(ctx context.Context) error {
+			var (
+				addonsMR = &resourcesv1alpha1.ManagedResource{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "addons",
+						Namespace: namespace.Name,
+					},
+				}
+				resourceManagerDeployment = &appsv1.Deployment{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      v1beta1constants.DeploymentNameGardenerResourceManager,
+						Namespace: namespace.Name,
+					},
+				}
+			)
+
+			if err := seedClient.Get(ctx, client.ObjectKeyFromObject(addonsMR), addonsMR); err != nil {
+				// If the MR is already gone, then nothing to do here
+				if apierrors.IsNotFound(err) {
+					return nil
+				}
+				return err
+			}
+
+			if err := seedClient.Get(ctx, client.ObjectKeyFromObject(resourceManagerDeployment), resourceManagerDeployment); err != nil {
+				return err
+			}
+
+			replicas := pointer.Int32Deref(resourceManagerDeployment.Spec.Replicas, 0)
+			if replicas > 0 {
+				patch := client.MergeFrom(resourceManagerDeployment.DeepCopy())
+				resourceManagerDeployment.Spec.Replicas = pointer.Int32(0)
+				if err := seedClient.Patch(ctx, resourceManagerDeployment, patch); err != nil {
+					return fmt.Errorf("failed to scale gardener-resource-manager deployment to zero %q: %w", client.ObjectKeyFromObject(resourceManagerDeployment), err)
+				}
+			}
+
+			patch := client.MergeFrom(addonsMR.DeepCopy())
+			addonsMR.Finalizers = nil
+			if err := seedClient.Patch(ctx, addonsMR, patch); err != nil {
+				return fmt.Errorf("failed to patch ManagedResource %q: %w", client.ObjectKeyFromObject(addonsMR), err)
+			}
+
+			if err := managedresources.DeleteForShoot(ctx, seedClient, namespace.Name, "addons"); err != nil {
+				return err
+			}
+
+			if err := managedresources.WaitUntilDeleted(ctx, seedClient, namespace.Name, "addons"); err != nil {
+				return err
+			}
```

Comment:
This waits forever. Also, why is it needed in the first place? We have removed all finalizers, so above delete call will make the resources to immediately disappear from etcd?